### PR TITLE
[FLINK-30816][sql-client] Fix SQL Client always uses the highest packaged version to connect to gateway

### PR DESCRIPTION
**Pull request by fsk119**
…kaged version to connect to gateway

## What is the purpose of the change

*When connecting to the gateway, the client should check that the gateway version satisfies the requirements and then open the session later.*

## Verifying this change

This change is already covered by existing tests, such as *ExecutorImplITCase#testConnectToEndpointWithV1Version*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/handler/util/GetApiVersionHandler.java [hunk 1]
```diff
@@ -40,12 +42,31 @@
...
             Map<String, String> responseHeaders,
             MessageHeaders<EmptyRequestBody, GetApiVersionResponseBody, EmptyMessageParameters>
                     messageHeaders) {
+        this(
+                service,
+                responseHeaders,
+                messageHeaders,
+                Arrays.stream(SqlGatewayRestAPIVersion.values())
+                        .filter(SqlGatewayRestAPIVersion::isStableVersion)
+                        .collect(Collectors.toList()));
```
Comment (LadyForest) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/handler/util/GetApiVersionHandler.java:58]: Nit: encapsulate the logic within `SqlGatewayRestAPIVersion` like `getStableVersions` or `getSupportedVersions`?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/ExecutorImpl.java [hunk 2]
```diff
@@ -429,4 +453,46 @@ private CompletableFuture<OperationStatusResponseBody> closeOperationAsync(
...
+        SqlGatewayRestAPIVersion clientVersion = SqlGatewayRestAPIVersion.getDefaultVersion();
+
+        if (gatewayVersions.contains(clientVersion)) {
+            return clientVersion;
+        } else {
+            SqlGatewayRestAPIVersion latestVersion =
+                    RestAPIVersion.getLatestVersion(gatewayVersions);
+            if (latestVersion.equals(SqlGatewayRestAPIVersion.V1)) {
+                throw new SqlExecutionException(
+                        "Currently SQL Client only supports to connect to the REST endpoint whose API version is larger than V1.");
```
Comment (LadyForest) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/ExecutorImpl.java:493]: Nit: suggested by Grammarly
`Currently SQL Client only supports to connect to the REST endpoint whose API version is larger than V1.` 
 => 
`Currently, SQL Client only supports connecting to the REST endpoint with API version larger than V1.`

**Review by LadyForest — COMMENTED** [review]
Thanks for the contribution! The PR looks good, and I just left some minor comments.

**Comment by fsk119** [comment]
@flinkbot run azure

**Comment by fsk119** [comment]
In my private pipeline, all tests pass. 

https://dev.azure.com/1059623455/Flink/_build/results?buildId=518&view=results

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/ExecutorImpl.java [hunk 3]
```diff
@@ -429,4 +453,46 @@ private CompletableFuture<OperationStatusResponseBody> closeOperationAsync(
...
+                                        // to build the target URL without API version.
+                                        Collections.min(
+                                                Arrays.stream(SqlGatewayRestAPIVersion.values())
+                                                        .filter(
+                                                                SqlGatewayRestAPIVersion
+                                                                        ::isStableVersion)
+                                                        .collect(Collectors.toList()))))
+                        .getVersions().stream()
+                        .map(SqlGatewayRestAPIVersion::valueOf)
+                        .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (fsk119) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/ExecutorImpl.java]: Considering current implementation is a little hacky, I think it's better we can keep it inside. I think others will not use `getInitialStableVersion`.

**Review by LadyForest — APPROVED** [review]
Thanks for the update, LGTM

**Pull request closed without merging**